kernel/time/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+/// is larger than that of the previous source.
+///
+/// Returns whether the previous source was overwritten.
+pub fn register_clock_source<T>(frequency: u64) -> bool
+where
+    T: ClockSource,
+{
+    let old_frequency = T::ClockType::frequency_atomic().load(Ordering::SeqCst);
+    if frequency > old_frequency {
+        let now_fn = T::ClockType::now_fn();
+        now_fn.store(T::now);
+
+        T::ClockType::store_unit_to_duration_func(T::unit_to_duration);
+        T::ClockType::store_duration_to_unit_func(T::duration_to_unit);
+
+        T::ClockType::frequency_atomic().store(frequency, Ordering::SeqCst);
+
+        true
+    } else {
+        false
+    }
+}
+
+/// Returns the current time.
+///
+/// Monotonic clocks return an [`Instant`] whereas realtime clocks return a
+/// [`Duration`] signifying the time since 12:00am January 1st 1970 (i.e. Unix
+/// time).
+///
+/// # Panics
+///
+/// This function will panic if called prior to registering a clock using
+/// [`register_clock_source`]. [`register_clock_source`] must return [`Ok`] and
+/// the [`ClockType`] of the registered clock must be the same as `T`.
+pub fn now<T>() -> T::Unit
+where
+    T: ClockType,
+{
+    let f = T::now_fn().load();
+    f()
+}
+
+/// A clock source.
+pub trait ClockSource {
+    /// The type of clock (either [`Monotonic`] or [`Realtime`]).
+    type ClockType: ClockType;
+
+    /// The current time according to the clock.
+    ///
+    /// For monotonic clocks this is usually the time since boot, and for
+    /// realtime clocks it's the time since 12:00am January 1st 1970 (i.e.
+    /// Unix time).
+    fn now() -> <Self::ClockType as ClockType>::Unit;
+
+    /// Converts a [`ClockType::Unit`] into a [`Duration`].
+    ///
+    /// Monotonic clocks should just return the [`ClockType::Unit`].
+    fn unit_to_duration(unit: <Self::ClockType as ClockType>::Unit) -> Duration;
+
+    /// Converts a [`Duration`] into a [`ClockType::Unit`].
+    ///
+    /// Monotonic clocks should just return the [`Duration`].
+    fn duration_to_unit(duration: Duration) -> <Self::ClockType as ClockType>::Unit;
+}
+
+/// A hardware clock that can sleep without relying on interrupts.
+pub trait EarlySleeper: ClockSource<ClockType = Monotonic> {
+    /// Wait for the given `duration`.
+    ///
+    /// This function spins the current task rather than sleeping it and so,
+    /// when possible, the `sleep` crate should be used.
+    ///
+    /// However, unlike the `sleep` crate, this function doesn't rely on
+    /// interrupts, and can be used prior to the scheduler being initiated.
+    ///
+    /// # Note to Implementors
+    ///
+    /// The default implementation of this function uses [`ClockSource::now`] -
+    /// it can only be used if [`ClockSource::now`] doesn't rely on
+    /// interrupts.
+    fn sleep(duration: Duration) {
+        let start = Self::now();
+        while Self::now() < start + duration {}
```

Comment:
overflow check (?)

Reply:
the overflow check is now done in the addition